### ISSUE TITLE
keyboard light turns on and doesnt turn off after plugging in charger when lid is closed and in sleep mode (proart px13)

### ISSUE THREAD

**Issue by thatonehhdev** [seerge/g-helper#3335]
### Rules

- [X] I made myself familiar with the <a href='https://github.com/seerge/g-helper?tab=readme-ov-file'>Readme</a>, <a href='https://github.com/seerge/g-helper/wiki/FAQ'>FAQ</a> and <a href='https://github.com/seerge/g-helper/wiki/Troubleshooting'>Troubleshooting</a>.
- [X] I understand that, if insufficient information or no app logs will be provided, my issue will be closed without an answer.

### What's wrong?

when i close the lid of my proart px13, the keyboard light turns off as expected. but when i plug in a charger, the light turns on again. if i unplug the charger, the light briefly turns off and then turns back on again. it never turns off even if i set a timeout for the light to turn off. it doesnt seem to have that problem when i dont use ghelper.

### How to reproduce the bug?

close lid and plug in charger, then unplug it.

### Logs

[log.txt](https://github.com/user-attachments/files/17606070/log.txt)

### Device and Model

asus proart px13

### Armoury Crate

Uninstalled

### Asus Services

none

### Version

0.197.0

### OS

windows 11 24h2

**Comment by seerge** [comment]
@thatonehhdev hello,

When device is in sleep all processes (including G-Helper) are suspended by system, so it can't neither turn on nor turn off backlight. It means this is how firmware is programmed on your device.

Keep in mind, that when laptop is in sleep - keyboard has a dedicated sleep backlight mode (blinking). Try to turn it off completely form `Extra` settings by unchecking corresponding `Sleep` checkbox.

Thanks

**Comment by thatonehhdev** [comment]
ah ok, so its the firmware, i already unchecked the sleep box tho.

**Comment by thatonehhdev** [comment]
but if it is the firmwares problem, why does it not happen when i use the asus services?

**Comment by seerge** [comment]
@thatonehhdev how exactly does the backlight look like ? Is it static or blinking ?

May be the laptop wasn't actually sleeping in that moment, but was just awake with closed lid ? Then when you plug in G-Helper would set last used performance mode used in plugged in state + set last used backlight brightness as well. So if you had backlight set to anything but OFF - it would set that state again.

**Comment by seerge** [comment]
@thatonehhdev I have trough about it a bit more, and have an idea what could happen.

If lid is closed but device is not sleeping (yet), i.e. all apps are still running. G-Helper would process plug-in effect as I explained in a previous message - it would set last used backlight levels for plugged-in state.

I have added one more check in the app, and when lid is closed it won't try to change backlight at all. Try this build in different scenarios and let me know if it makes a difference

[GHelper.zip](https://github.com/user-attachments/files/17609957/GHelper.zip)

**Comment by thatonehhdev** [comment]
yea i think that makes sense, it might be because of the "modern standby" thing replacing sleep mode in windows. now that you say that tho, it does make a lot more sense. it behaves the same as if the lid is open, it momentairly switches off and back on again. same as if the lid is open.

**Comment by seerge** [comment]
@thatonehhdev ok, but does the build I linked here fix your issue ?

**Comment by thatonehhdev** [comment]
it seems to have fixed it

**Comment by thatonehhdev** [comment]
it doesnt turn on the kb light anymore when the lid is closed, so i think its fixed

**Comment by thatonehhdev** [comment]
i do need more time to test it tho

**Comment by seerge** [comment]
@thatonehhdev ok, then i'm closing it as resolved. Feel free to reopen / post an update if you will spot something. Thanks

**Comment by thatonehhdev** [comment]
ok thanks :)

**Comment by thatonehhdev** [comment]
the kb light turned back on after i left it plugged in for a few minutes, but unplugging the charger and plugging it in again doesnt make the kb light momentairly turn off, so i dont think ghelper is running.

**Comment by thatonehhdev** [comment]
i think its all due to the modern standby, but i havent found a way to disable it. i do wonder if there is a way to keep ghelper running even in modern standby, so some kind of priority maybe.

**Comment by seerge** [comment]
@thatonehhdev i didn't completely get you. So keyboard turned ON even without G-Helper running ?

**Comment by thatonehhdev** [comment]
what i am thinking is that modern standby suspended ghelper and then the kb light somehow turned on. i am checking how to turn off modern standby and enable legacy sleep mode but i dont think thats possible, it seems to me a firmware thing.

**Comment by seerge** [comment]
@thatonehhdev but how did it turn on ? Even if G-Helper would be running - it's not "holding" backlight anyhow :) 

What it does is monitors plug / unplug events and then just sets last used backlight level on that moment. After - it does nothing (and it shouldn't do anything as well)

**Comment by thatonehhdev** [comment]
yea i am confused too, the reason why i am thinking it is modern standbys problem is because it was fine for 10 minutes, then the light turned on

**Comment by seerge** [comment]
@thatonehhdev check G-Helper log for exact that moment of time, every single action that G-Helper does it also logs.

If there is nothing - then it's something else that has turned light on (leftovers from asus apps?) or firmware ?

**Comment by thatonehhdev** [comment]
[log.txt](https://github.com/user-attachments/files/17611851/log.txt)
heres the logs if you want them. is the TUF Brightness the kb light?

**Comment by thatonehhdev** [comment]
i dont know the exact time

**Comment by seerge** [comment]
@thatonehhdev yes and so is ``19B6|64: 5A-BA-C5-C4-00`` (last byte is backlight level). 

```
11/4/2024 3:46:05 AM: TUF Brightness = 128 : OK
11/4/2024 3:46:05 AM: Lid 19B6|64: 5A-BA-C5-C4-00
```

But at what timestamp should i look at ? when did brightness turned on ?

Also important note : G-Helper need to be running **before** you close the lid, so it would know that lid is closed. 

How do you even see that backlight turns on when lid is closed btw ?

**Comment by thatonehhdev** [comment]
i can see it through the cooling vents

**Comment by thatonehhdev** [comment]
yea i have it running. i might need to spend some time to recreate the issue

**Comment by thatonehhdev** [comment]
i think it happened between 2:18 and 2:34

**Comment by thatonehhdev** [comment]
because after that i opened the lid and closed it again to turn off the kb light

**Comment by thatonehhdev** [comment]
what is it doing at 34


11/4/2024 3:18:29 AM: BatteryLimit = 100 : OK
11/4/2024 3:18:29 AM: Lid Open
11/4/2024 3:18:29 AM: TUF Brightness = 129 : OK
11/4/2024 3:18:29 AM: Lid 19B6|64: 5A-BA-C5-C4-01
11/4/2024 3:18:29 AM: Eco flag : 1
11/4/2024 3:18:29 AM: Mux flag : -65536
11/4/2024 3:18:29 AM: \\?\DISPLAY#SDC41B0#5&32e93a4f&2&UID256#{e6f07b5f-ee97-4a90-b076-33f57bf4eaa7} DISPLAYCONFIG_OUTPUT_TECHNOLOGY_INTERNAL
11/4/2024 3:18:29 AM: \\?\DISPLAY#SDC41B0#5&32e93a4f&2&UID256#{e6f07b5f-ee97-4a90-b076-33f57bf4eaa7} \\.\DISPLAY1\Monitor0
11/4/2024 3:18:29 AM: ATNA33AA08-0  HDR: False 10 1 False
11/4/2024 3:18:30 AM: PowerLimit A3 = 15 : OK
11/4/2024 3:18:30 AM: PowerLimit A0 = 15 : OK
11/4/2024 3:18:30 AM: PowerLimit C1 = 15 : OK
11/4/2024 3:18:31 AM: UV: -40 OK
11/4/2024 3:18:31 AM: CPU Temp: 90 OK
11/4/2024 3:34:02 AM: Forcing Standard Mode on shutdown / hibernation
11/4/2024 3:34:02 AM: GPUEco = 0 : OK
11/4/2024 3:34:02 AM: BatteryLimit = 100 : OK
11/4/2024 3:42:16 AM: ------------
11/4/2024 3:42:17 AM: App launched: ProArt PX13 HN7306WV_HN7306WV :0.198.0.0en-US.
11/4/2024 3:42:17 AM: Start Count: 71

**Comment by thatonehhdev** [comment]
im not sure about asus apps turning it on, the only one i have is dial pad control panel. but there could be services running, ghelper shows none tho.

**Comment by thatonehhdev** [comment]
![image](https://github.com/user-attachments/assets/79b563b2-f5ba-4d9e-a31c-02d1361f7245)
these are the asus services. its late here(you can tell from the logs) so i need to go to sleep. when i have time, i will try to recreate the issue.

**Comment by seerge** [comment]
@thatonehhdev ok, I have added a bit more verbose logging (but didn't change anything else)

Can you try this build, and post a log if it happens again (ideally with approximate timestamp when it happened)

[GHelper.zip](https://github.com/user-attachments/files/17611961/GHelper.zip)

**Comment by thatonehhdev** [comment]
i managed to recreate the issue. it doesnt turn on by it self, it turns on if you leave it for some time and then plug it in. im not sure if it is ghelpers problem then
[log.txt](https://github.com/user-attachments/files/17612019/log.txt)

**Comment by thatonehhdev** [comment]
also just a question, what is the aurora mode?

**Comment by seerge** [comment]
@thatonehhdev 

In this log fragment there is not a single backlight command run by G-Helper between lid close
```
11/4/2024 4:18:08 AM: Lid Closed
```
and lid open
```
11/4/2024 4:45:10 AM: Lid Open
```

So if something turned backlight on while lid was closed - it wasn't G-Helper (could be firmware or something else).

---

> also just a question, what is the aurora mode? is it visual mode?

It's how backlight modes (i.e. Static, Strobbing, Breathing, etc) are called in Armoury Crate. G-Helper was initially crated for ROG / TUF devices - and that's why many features are mainly called like  in Armoury that handles those models  internally :)

**Comment by thatonehhdev** [comment]
I will try to recreate it while using Asus services. if I can, I will report it to Asus.

**Comment by thatonehhdev** [comment]
but the check to see if the lid is closed does help a lot

**Comment by thatonehhdev** [comment]
i cant recreate it with asus services. when i run the service called asus optimisation, it solves the problem but the brightness adjustment gets all messed up. is there any way to prevent ghelper from controlling the kb light and use the asus service for controlling the kb light only?

**Comment by seerge** [comment]
@thatonehhdev when Optimization service is running G-Helper won't handle backlight controls. 

If you start / stop asus services within `Extra` settings - G-Helper is immediately aware of that, if you do it manually via windows services - you need to restart it.

**Comment by thatonehhdev** [comment]
can you add a check that goes like this: when charger plugged in, check if lid is closed. if closed, set backlight brightness to 0.
also i did it with windows services, the kb light adjustment works but the icon that pops up is messed up

**Comment by thatonehhdev** [comment]
ah wait, i tried it again and now its fine

**Comment by thatonehhdev** [comment]
i am a bit worried that asus optimisation might cause other problems

**Comment by seerge** [comment]
@thatonehhdev 
> can you add a check that goes like this: when charger plugged in, check if lid is closed. if closed, set backlight brightness to 0.

It's already like that :) It sets backlight to OFF when lid is closed. And doesn't touch it until it's open.

**Comment by thatonehhdev** [comment]
ok i solved the kb light issue. i just need asus optimisation running and then its fine. but i am worried that asus optimisation might cause other problems. also after a restart, the icons dont pop up anymore for fn lock and other shortcuts. i know its because ghelper is not in charge of that anymore because i enabled asus optimisations, but is there any way to get it back?
there are also some other minor bugs i found, for example the dgpu sometimes enables it self after a restart and fn lock gets messed up after hibernation. i am aware that the dgpu thing is because of the nvidia driver, but can you add something that actively checks if the dgpu is in the state its supposed to be or was last set to?

**Comment by thatonehhdev** [comment]
i can solve the dgpu issue by disabling the nvidia driver, but it gets annoying as i need to restart my laptop if i need to use the dgpu and then restart it again after i finish using it (i disabled the kernel mode driver in autoruns)

**Comment by seerge** [comment]
@thatonehhdev
> but i am worried that asus optimisation might cause other problems. also after a restart, the icons dont pop up anymore for fn lock and other shortcuts. i know its because ghelper is not in charge of that anymore because i enabled asus optimisations, but is there any way to get it back?

If you mean G-Helper's own notifications, then it obviously won't show them as it doesn't process hotkeys :) And it's not aware of FN-Lock status set by Optimization service.

> there are also some other minor bugs i found, for example the dgpu sometimes enables it self after a restart and fn lock gets messed up after hibernation. 

What exactly do you mean under dGPU enables ? Does it switch from Eco to Standard ? Do you have `Enable GPU on shutdown` checked under `Extra`?

> the fn lock issue is after hibernation, it thinks the fn lock is on, but its not and i need to turn it off and back on for it to actually turn on.

If Optimization service is running - G-Helper won't handle FN-Lock in order to not to interfere.

**Comment by seerge** [comment]
@thatonehhdev as for you backlight behavior and lid being closed. Does backlight turn OFF on lid close when you don't have G-Helper running at all, but you have asus services running ? (i.e. do they also "handle" this somehow?)

**Comment by thatonehhdev** [comment]
everything with the kb backlight works fine with asus services.

yes it switches from eco to standard, i have gpu on shutdown enabled.

**Comment by thatonehhdev** [comment]
and when i dont use asus services, the fn lock turns off by it self after hibernation. but ghelper doesnt know that fn lock turned off. so maybe add something that sets fn lock to last state after startup

**Comment by thatonehhdev** [comment]
it might not be problems after a restart with the dgpu, it might also be hibernation. all i know is i had that problem before. i will try to recreate it..

**Comment by thatonehhdev** [comment]
the camera on and off also gets messed up by hibernation

**Comment by seerge** [comment]
@thatonehhdev 
> yes it switches from eco to standard, i have gpu on shutdown enabled.

Well, it enables GPU intentionally. As you have literally selected this option. This thing exists only to prevent some [nvidia driver issues that can happen](https://github.com/seerge/g-helper/discussions/1042). If you don't have those issues - you may uncheck it, and then GPU will stay disabled.

>and when i dont use asus services, the fn lock turns off by it self after hibernation. but ghelper doesnt know that fn lock turned off. so maybe add something that sets fn lock to last state after startup

I didn't get it completely tbh. On each start / wake up (incl hibernation) G-Helper would set last used FN-Lock status intentionally. You can see it in the log lines
```
11/2/2024 3:44:23 PM: FnLock = 1 : OK
11/2/2024 3:44:23 PM: USB FnLock 19B6|64: 5A-D0-4E-00
```

It looks more like Asus service is also toggling it then ?

**Comment by thatonehhdev** [comment]
huh thats weird, i have all asus services disabled. i cant seem to recreate the dgpu problem so maybe it was just me remembering weird stuff.

**Comment by seerge** [comment]
@thatonehhdev I have one more idea to test for your lid and backlight.

Try this build (stop all asus services as well) and check if backlight stays off when lid is closed.
[GHelper.zip](https://github.com/user-attachments/files/17616312/GHelper.zip)

**Comment by thatonehhdev** [comment]
ok let me try it. it takes some time to test because i need to wait for windows modern standby to do its thing(20 minutes)

**Comment by thatonehhdev** [comment]
i need to also do more research for how to disable modern standby. i think its causing all of the problems.

**Comment by thatonehhdev** [comment]
is there any way to "install" ghelper so windows settings can see it in the apps section?

**Comment by thatonehhdev** [comment]
ok, i made a msi installer that adds it to the windows list of apps(required to run in modern standby.

**Comment by thatonehhdev** [comment]
or maybe its because of this. the laptop suspends ghelper. i unplug the charger and plug it in. wakes the laptop and puts it in some kind of low power mode between on and modern standby. ghelper doesnt start. kb light stays on.

**Comment by seerge** [comment]
@thatonehhdev do you have any kind dedicated setting for backlight on lid close in native Asus apps? (myasus / proart , etc)?

**Comment by thatonehhdev** [comment]
no, i cant find any.

**Comment by seerge** [comment]
@thatonehhdev cause even if G-Helper won't be suspended in sleep - it will not "prevent" backlight from turning ON :) I was thinking if there is some hidden USB command that Asus Services set to keyboard to stop it from turning backlight on or so ...

**Comment by seerge** [comment]
@thatonehhdev can you do one simple test.

1. Turn all Asus services off
2. In G-Helper set backlight level to ON with hotkeys
3. Open `Extra` settings and uncheck `Awake` state there. Does backlight turn OFF ?
4. If it does - put laptop to sleep and monitor if it stays off

**Comment by thatonehhdev** [comment]
yes, it stays off then.

**Comment by thatonehhdev** [comment]
so i guess adding a function that disables the kb light in all conditions when the lid closes will solve it.

**Comment by thatonehhdev** [comment]
i think it worked, i need to test it more thoroughly tho.

**Comment by thatonehhdev** [comment]
also is there any way to undervolt more than 40?

**Comment by seerge** [comment]
@thatonehhdev 

> i think it worked, i need to test it more thoroughly tho.


Ok, great. Can you check if this build still works as you expect ? (i did some cleanup with backlight logic)
[GHelper.zip](https://github.com/user-attachments/files/17659524/GHelper.zip)

> also is there any way to undervolt more than 40?

You can do that, [check power user settings](https://github.com/seerge/g-helper/wiki/Power-user-settings#custom-cpu-undervolting-limit)

**Comment by thatonehhdev** [comment]
ok, i will test it.

**Comment by thatonehhdev** [comment]
yes, it works. the keyboard light doesnt turn on with the lid closed. thanks :)

**Comment by seerge** [comment]
@thatonehhdev ok, amazing. Then I will include this in the next release. Thanks for checking.
